<a href="https://colab.research.google.com/github/wanita-buike/Data-science-road-map/blob/latest_branch/ASSIGNMENT_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

# Loading the Dataset

In [ ]:
df = pd.read_csv ("/content/housing_dataset.csv")

## SECTION 1. Data Exploration

### Question 1 : Displaying the first five rows

In [ ]:
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15.0,5612.0,1283.0,1015.0,472.0,1.4936,66900.0
1,-114.47,34.40,19.0,7650.0,1901.0,1129.0,463.0,1.8200,80100.0
2,-114.56,33.69,17.0,720.0,174.0,333.0,117.0,1.6509,85700.0
3,-114.57,33.64,14.0,1501.0,337.0,515.0,226.0,3.1917,73400.0
4,-114.57,33.57,20.0,1454.0,326.0,624.0,262.0,1.9250,65500.0


### Question 2 : Checking the shape of the Data Frame

In [ ]:
df.shape # rows, columns

(17020, 9)

Text Cell explaning the output Of Question 2

The result from df.shape shows the number of rows and columns in the DataFrame.

The first number represents the number of rows (of housing) in the dataset,

The second number represents the number of columns ie; Longitiude, Latitiude, housing_median age, Total_rooms etc.

That is the output is (17020, 9), it means the dataset contains 17020 Housing_Dataset and 9 columns.



### Question 3: Checking the Dataset Information

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17020 entries, 0 to 17019
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           17020 non-null  float64
 1   latitude            17020 non-null  float64
 2   housing_median_age  17020 non-null  float64
 3   total_rooms         17020 non-null  float64
 4   total_bedrooms      16990 non-null  float64
 5   population          16980 non-null  float64
 6   households          17020 non-null  float64
 7   median_income       16990 non-null  float64
 8   median_house_value  17020 non-null  float64
dtypes: float64(9)
memory usage: 1.2 MB


### Question 4: Checking the Statistical Summary

In [ ]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
longitude,17020.0,-119.562066,2.005157,-124.3500,-121.7900,-118.49000,-118.0000,-114.3100
latitude,17020.0,35.625785,2.137446,32.5400,33.9300,34.25000,37.7200,41.9500
housing_median_age,17020.0,28.590541,12.588782,1.0000,18.0000,29.00000,37.0000,52.0000
total_rooms,17020.0,2864.361927,9385.577088,2.0000,1462.0000,2127.50000,3153.0000,379370.0000
total_bedrooms,16990.0,539.285756,421.598465,1.0000,296.0000,434.00000,648.7500,6445.0000
population,16980.0,1429.529505,1148.157282,3.0000,789.0000,1166.50000,1721.0000,35682.0000
households,17020.0,501.103408,384.400904,1.0000,282.0000,409.00000,605.2500,6082.0000
median_income,16990.0,3.883400,1.908640,0.4999,2.5659,3.54425,4.7676,15.0001
median_house_value,17020.0,205747.585488,118642.283445,-500001.0000,118800.0000,179800.00000,264700.0000,500001.0000


#####  Observation of Question 4

The columns are on different scale, For instance, the median_house value has much larger scale than the median_income and latitiude, which suggests that the data set may need scaling before modeling.

### Question 5 : Check for Missing  Values

In [ ]:
df.isnull().sum()

,0
longitude,0
latitude,0
housing_median_age,0
total_rooms,0
total_bedrooms,30
population,40
households,0
median_income,30
median_house_value,0


The total_bedrooms column has missing values which is 207, while the other columns do not contain missing data. This means the dataset will require handling missing values, such as filling them with the mean or dropping the rows.

### Question 6 : Checking for Duplicate

In [ ]:
df.duplicated().sum()

np.int64(20)

## SECTION 2: Data Cleaning

### Question 7: Fix missing values

### 7.1 : Fill with Median

In [ ]:
total_bedrooms_median = df['total_bedrooms'].median()
print(total_bedrooms_median)

434.0


In [ ]:
df['total_bedrooms'] = df['total_bedrooms'].fillna(total_bedrooms_median) # Fill with median because of outliers

In [ ]:
df['total_bedrooms'].isnull().sum()

np.int64(0)

In [ ]:
population_median = df['population'].median()
print(population_median)

1166.5


In [ ]:
df['population'] = df['population'].fillna(population_median) # Fill with median because of outliers

In [ ]:
df['population'].isnull().sum()

np.int64(0)

### 7.2: Fill with Mean

In [ ]:
income_mean = df['median_income'].mean()
print(income_mean)

3.883399793996468


In [ ]:
df['median_income'] = df['median_income'].fillna(income_mean) # Fill with median because of outliers

In [ ]:
df['median_income'].isnull().sum() # Sanity check

np.int64(0)

#### a :What did i choose to do

Columns i handled are ; total_bedrooms , population , median_income

so, i had to fill the total_bedrooms and population with the Median while i had to fill the median_income with the mean.

### b ; Why did you make that choice? Think about the distribution of the column and whether there are outliers.

why i made that choice is that, total_bedrooms and population contain outliers, and the mean is safer,while median_income is usually more evenly distributed. so mean works well.

### C ; Do a sanity check to confirm there are no more missing values in that column

I did a Sanity Check After filling the missing values, I ran df.isnull().sum() and confirmed that all columns now have missing values which is 0.

### Question 8: Handling Duplicates


In [ ]:
df.shape # Before dropping the duplicates

(17020, 9)

In [ ]:
df = df.drop_duplicates()

In [ ]:
df.shape # After dropping the duplicates

(17000, 9)

### **Text cell Explanation**

Explain why keeping duplicate rows in a dataset can be a problem, especially when you are training a machine learning model.
            
Duplicate rows can cause problems in a dataset because it reduces the accuracy of predictions and lead to overfitting, where the model performs well on training data but poorly on new, unseen data. This can make a machine learning model biased, since the model may learn patterns from duplicated data more than it should.

Removing duplicates helps ensure that the dataset is clean, balanced, and more reliable for training a machine learning model.

## SECTION 3 ; Feature Scaling

### Question 9

### What is Feature Scaling and Why is it Necessary?

Feature scaling is the transforming of numerical data into the same range, usually between 0 and 1.

Feature scaling is necessary because many machine learning algorithms such as Linear Regression, and Gradient Descent-based models perform better and faster when data is scaled.

### What Problem Would Arise Without Scaling?

If the housing dataset is used without scaling, features with large values like total_rooms or population will dominate features with smaller values like median_income which wil cause the model to learn incorrectly and produce poor predictions.

### Question 10


In [ ]:
# Min-Max scaling on a single column

In [ ]:
cols_to_scale = ['housing_median_age', 'total_rooms', 'median_income', 'median_house_value']

In [ ]:
for col in cols_to_scale:
    col_min = df[col].min()
    col_max = df[col].max()
    df[col + '_scaled'] = (df[col] - col_min) / (col_max - col_min)


### Question 11

In [ ]:
scaled_cols = [
    'housing_median_age_scaled',
    'total_rooms_scaled',
    'median_income_scaled',
    'median_house_value_scaled'
]

for col in scaled_cols:
    print(f"{col}: min = {df[col].min():.1f}, max = {df[col].max():.1f}")


housing_median_age_scaled: min = 0.0, max = 1.0
total_rooms_scaled: min = 0.0, max = 1.0
median_income_scaled: min = 0.0, max = 1.0
median_house_value_scaled: min = 0.0, max = 1.0


### Question 12


In [ ]:
df[['median_income', 'median_income_scaled']].head(10)


,median_income,median_income_scaled
0,1.4936,0.068530
1,1.8200,0.091040
2,1.6509,0.079378
3,3.1917,0.185639
4,1.9250,0.098281
5,3.3438,0.196128
6,2.6768,0.150129
7,1.7083,0.083337
8,2.1782,0.115743
9,2.1908,0.116612


### Text Cell Explanation (Describe what you notice about the relationship between the original values and the scaled values).

Looking at median_income and median_income_scaled side by side, the relative
ordering of the values is completely preserved — a row with a higher income still
has a higher scaled value. The only thing that changed is the range. The original
values sit somewhere between 0.5 and 15, while the scaled values are compressed
into the range 0 to 1. The row with the lowest income in the entire dataset gets
a scaled value of exactly 0.0, and the row with the highest income gets exactly
1.0. Every other row falls proportionally in between. No information about
relative differences is lost — the scale is simply standardised.